src/mdns/mdns_mdnssd.cpp

```diff
@@ -361,331 +428,203 @@ void PublisherMDnsSd::HandleServiceRegisterResult(DNSServiceRef         aService
-                                               DNSServiceErrorType aErrorCode,
+                                               DNSServiceErrorType aError,
                                                void *              aContext)
 {
-    static_cast<PublisherMDnsSd *>(aContext)->HandleRegisterHostResult(aHostsConnection, aHostRecord, aFlags,
-                                                                       aErrorCode);
+    static_cast<PublisherMDnsSd *>(aContext)->HandleRegisterHostResult(aServiceRef, aRecordRef, aFlags, aError);
 }
 
-void PublisherMDnsSd::HandleRegisterHostResult(DNSServiceRef       aHostsConnection,
-                                               DNSRecordRef        aHostRecord,
+void PublisherMDnsSd::HandleRegisterHostResult(DNSServiceRef       aServiceRef,
+                                               DNSRecordRef        aRecordRef,
                                                DNSServiceFlags     aFlags,
-                                               DNSServiceErrorType aErrorCode)
+                                               DNSServiceErrorType aError)
 {
-    OTBR_UNUSED_VARIABLE(aHostsConnection);
     OTBR_UNUSED_VARIABLE(aFlags);
 
-    HostIterator host = FindPublishedHost(aHostRecord);
-    std::string  hostName;
+    otbrError           error   = DNSErrorToOtbrError(aError);
+    HostRegistrationPtr hostReg = FindHostRegistration(aServiceRef, aRecordRef);
+
+    std::string hostName;
+
+    VerifyOrExit(hostReg != nullptr);
 
-    VerifyOrExit(host != mHosts.end());
-    hostName = host->mName;
+    hostName = MakeFullHostName(hostReg->mName);
 
     otbrLogInfo("Received reply for host %s", hostName.c_str());
 
-    if (aErrorCode == kDNSServiceErr_NoError)
+    if (error == OTBR_ERROR_NONE)
     {
         otbrLogInfo("Successfully registered host %s", hostName.c_str());
+        hostReg->Complete(OTBR_ERROR_NONE);
     }
     else
     {
-        otbrLogWarning("failed to register host %s for mdnssd error: %s", hostName.c_str(),
-                       DNSErrorToString(aErrorCode));
-
-        DiscardHost(hostName, /* aSendGoodbye */ false);
-    }
-
-    if (mHostHandler != nullptr)
-    {
-        mHostHandler(hostName, DNSErrorToOtbrError(aErrorCode), mHostHandlerContext);
+        otbrLogWarning("failed to register host %s for mdnssd error: %s", hostName.c_str(), DNSErrorToString(aError));
+        hostReg->Complete(error);
+        RemoveHostRegistration(hostReg->mName);
     }
 
 exit:
     return;
 }
 
-std::string PublisherMDnsSd::MakeFullName(const std::string &aName)
-{
-    return aName + "." + kDomain;
-}
-
 // See `regtype` parameter of the DNSServiceRegister() function for more information.
-std::string PublisherMDnsSd::MakeRegType(const std::string &aType, const SubTypeList &aSubTypeList)
+std::string PublisherMDnsSd::MakeRegType(const std::string &aType, SubTypeList aSubTypeList)
 {
     std::string regType = aType;
 
+    std::sort(aSubTypeList.begin(), aSubTypeList.end());
```

Comment:
Isn't aSubTypeList already sorted?

Reply:
We don't have the hard requirement on the parameter `aSubTypeList` that it must be sorted.